src/features/login/pages/LoginPage/index.tsx

```diff
@@ -15,50 +14,110 @@ import {
   ErrorMessage,
 } from './styled'
 
+const validators = {
+  email: (value: string) => {
+    if (typeof value !== 'string') return 'Invalid e-mail value type'
+    if (!value) return 'E-mail is required'
+    if (!/^[^\s@]+@[^\s@]+\.[^\s@]+$/u.test(value)) return 'Invalid e-mail'
+  },
+
+  password: (value: string) => {
+    if (typeof value !== 'string') return 'Invalid password value type'
+    if (!value) return 'Password is required'
+  },
+}
+
```

Comment:
Here you may be confused about a couple of things:

**1. Why an object?**
We could, in theory, make each of these functions as top-level variables. But, having them declared as properties of an object gives us a couple of benefits:

1. Simplified naming, as "email" is too generic, and "emailValidator" is not as simple anymore.
2. `validators` becomes a map, which allow use to:
   1. Dynamically access a validator (e.g.: `validators[inputName](inputValue)`)
   2. Iterate over validators

**2. Why outside the component?**
Generally speaking, anything that isn't dependent on props or state _should_ live outside of the render cycle. This ensures, for instance, that these values never change, and thus memoization just becomes easier without the need of `useMemo` or `useCallback` all around. Besides that, you get better testability and overall simplification of the context where these assets are used.

**3. Why the `IF`s and the `undefined` return**
It's a common and generally nice pattern for value validators to return either a string explaining the failure, or `undefined` representing no issue is found. This way, any check on the validation result becomes as simple as this:

```ts
if (validator.email(value)) { /* handle error */ }
```

Reply:
This is pretty clever! Damn, I wish I'd come up with this :D... With a combination of if statements it's super readable.